luci-app-yggdrasil: Add application for Yggdrasil

zhoreeq opened the pull request:
Signed-off-by: George Iv <57254463+zhoreeq@users.noreply.github.com>

Tested on: snapshot with yggdrasil v0.3.11-4

aparcar commented:
I think LuCI is moving on the JavaScript, from what I understand is the Lua to JavaScript conversion pretty straight forward. Would that be an option for you?

zhoreeq commented:
@aparcar sure, if there is API documentation. This app doesn't seem too complex to convert, but I can only see Lua API in the current docs and guidelines. 

If Lua apps are not totally deprecated yet, can it be converted after we add it?

aparcar commented:
I think the current documentation is [here](https://openwrt.github.io/luci/jsapi/) and [here](https://github.com/openwrt/luci/pull/3344/files) is a recent example of sombody moving from lua to js

zhoreeq commented:
@aparcar pushed a JS version. Review please.

wfleurant commented:
the transition to JS looks too good to be true

aparcar commented:
@wfleurant Thank you for looking into that. Could you please test the app and approve it? I'm then happy to merge.

aparcar commented:
@zhoreeq Ideally, not necessarily now, the Lua parts would be replaced by `rpcd` calls.

zhoreeq commented:
I've just removed luasocket dependency and controller is now using `yggdrasilctl -json` calls to get node stats. Unnecessary Lua code is also removed.

jow- commented on applications/luci-app-yggdrasil/htdocs/luci-static/resources/view/yggdrasil/keys.js:
```diff
@@ -0,0 +1,24 @@
+'use strict';
+'require fs';
+'require form';
+'require tools.widgets as widgets';
```
You don't need this require

jow- commented on applications/luci-app-yggdrasil/htdocs/luci-static/resources/view/yggdrasil/keys.js:
```diff
@@ -0,0 +1,24 @@
+'use strict';
+'require fs';
```
You don't need this require

jow- commented on applications/luci-app-yggdrasil/htdocs/luci-static/resources/view/yggdrasil/peers.js:
```diff
@@ -0,0 +1,33 @@
+'use strict';
+'require fs';
```
You don't need this require

jow- commented on applications/luci-app-yggdrasil/htdocs/luci-static/resources/view/yggdrasil/peers.js:
```diff
@@ -0,0 +1,33 @@
+'use strict';
+'require fs';
+'require form';
+'require tools.widgets as widgets';
```
You don't need this require

jow- commented on applications/luci-app-yggdrasil/htdocs/luci-static/resources/view/yggdrasil/session_firewall.js:
```diff
@@ -0,0 +1,40 @@
+'use strict';
+'require fs';
```
You don't need this require

jow- commented on applications/luci-app-yggdrasil/htdocs/luci-static/resources/view/yggdrasil/session_firewall.js:
```diff
@@ -0,0 +1,40 @@
+'use strict';
+'require fs';
+'require form';
+'require tools.widgets as widgets';
```
You don't need this require

jow- commented on applications/luci-app-yggdrasil/htdocs/luci-static/resources/view/yggdrasil/settings.js:
```diff
@@ -0,0 +1,59 @@
+'use strict';
+'require fs';
```
You don't need this require

jow- commented on applications/luci-app-yggdrasil/htdocs/luci-static/resources/view/yggdrasil/settings.js:
```diff
@@ -0,0 +1,59 @@
+'use strict';
+'require fs';
+'require form';
+'require tools.widgets as widgets';
```
You don't need this require

jow- commented on applications/luci-app-yggdrasil/htdocs/luci-static/resources/view/yggdrasil/status.js:
```diff
@@ -0,0 +1,113 @@
+'use strict';
+'require fs';
+'require form';
+'require tools.widgets as widgets';
```
You don't need this require

jow- commented on applications/luci-app-yggdrasil/htdocs/luci-static/resources/view/yggdrasil/tunnel_routing.js:
```diff
@@ -0,0 +1,51 @@
+'use strict';
+'require fs';
+'require form';
+'require tools.widgets as widgets';
```
You don't need this require

jow- commented on applications/luci-app-yggdrasil/htdocs/luci-static/resources/view/yggdrasil/tunnel_routing.js:
```diff
@@ -0,0 +1,51 @@
+'use strict';
+'require fs';
```
You don't need this require

jow- commented:
Apart from some redundant `require` statements, this looks quite good to me.

jow- commented on applications/luci-app-yggdrasil/root/usr/share/rpcd/acl.d/luci-app-yggdrasil.json:
```diff
@@ -0,0 +1,10 @@
+{
+        "luci-app-yggdrasil": {
+                "description": "Grant access to LuCI app yggdrasil",
+                "read": {
```
Since `yggdrasilctl` provides commands to modify the system state (such as `addPeer`, `removePeer`, `setTunTap` etc.) we should change `read` to `write` here. There will be no functional change in the LuCI app due to this, but it'll be needed in the future for proper permission handling.

zhoreeq replied:
Thanks for a review. Changed to write.

jow- commented on applications/luci-app-yggdrasil/htdocs/luci-static/resources/view/yggdrasil/status.js:
```diff
@@ -0,0 +1,96 @@
+	peerings.setAttribute("class", "table"); peerings.id = "yggdrasil-peerings";
+	var tr = document.createElement("tr");
+	tr.setAttribute("class", "tr table-titles");
+	["Endpoint", "Address", "Proto", "Uptime", "Rx / Tx"].forEach(function(t) {
```
Use `%1024mB'.format(bygtes)` instead.

jow- commented on applications/luci-app-yggdrasil/htdocs/luci-static/resources/view/yggdrasil/status.js:
```diff
@@ -0,0 +1,96 @@
+	peerings.setAttribute("class", "table"); peerings.id = "yggdrasil-peerings";
+	var tr = document.createElement("tr");
+	tr.setAttribute("class", "tr table-titles");
+	["Endpoint", "Address", "Proto", "Uptime", "Rx / Tx"].forEach(function(t) {
```
Maybe use `'%t'.format(9052.876473808)` which would result in something like `2h 30m 52s`

wfleurant replied:
i made these two format changes w/ @zhoreeq  https://github.com/zhoreeq/luci/pull/1

![image](https://user-images.githubusercontent.com/5460439/69893384-a3e5af80-12de-11ea-9f78-02e775ba9155.png)

zhoreeq replied:
It looks nice now :+1:

zhoreeq commented on applications/luci-app-yggdrasil/luasrc/controller/yggdrasil.lua:
```diff
@@ -0,0 +1,49 @@
+		return
+	end
+
+	entry({"admin", "services", "yggdrasil"}, firstchild(), "yggdrasil").dependent = true
```
:+1: That will make more sense. Should I change it to `{"admin", "network", "yggdrasil"}`? @wfleurant @aparcar @jow- @feckert

zhoreeq replied:
Since there are no objections, I guess that means "yes"

aparcar commented:
I'm not to familiar with the ACL, looks like you allow "exec" but not "stat", but at some point you use the latter. Is that a problem?

Also, the only lua bits are for creating a menu entry, isn't there a proper way to do so in JS?

zhoreeq commented:
> I'm not to familiar with the ACL, looks like you allow "exec" but not "stat", but at some point you use the latter. Is that a problem?

Doesn't seem to. I copied it from that samba app example you've provided, and it works without problems.

> Also, the only lua bits are for creating a menu entry, isn't there a proper way to do so in JS?

No idea. Maybe @jow- can tell?

jow- commented:
The `stat` permissions are implicitly granted by `luci-base` which permits stat on the entire fs.

Menu declarations require the Lua controller yet. Support for JSON-only menu declarations will follow later this week hopefully.

aparcar commented:
Okay so I'd merge this now and add a patch later for the json menu?

jow- commented:
Merged, thanks!
